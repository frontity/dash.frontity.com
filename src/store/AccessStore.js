import { types, flow, getParent } from 'mobx-state-tree';
import client, { setToken, removeToken } from '../graphql/client';

export default types
  .model({
    isWaiting: false,
    isLoggedIn: false,
    hasFailed: false,
  })
  .actions(self => ({
    logIn: flow(function* logIn({ value: emailAndPass }) {
      self.isWaiting = true;
      self.hasFailed = false;
      self.isLoggedIn = false;
      try {
        const { authenticateUser } = yield client.request(
          `
          mutation Signin($email: String!, $password: String!) {
            authenticateUser(email: $email, password: $password) {
              token
            }
          }
          `,
          emailAndPass,
        );

        setToken(authenticateUser.token);
        self.isLoggedIn = true;

        // Request user data
        const { user } = yield client.request(
          `
          query getUser {
            user {
              name
              email
            }
          }
          `,
        );

        console.log(user);

        getParent(self).setUser(user);

        // Redirects to main page
        getParent(self).router.openSitesPage();
      } catch (error) {
        console.log(error);
        self.isLoggedIn = false;
        self.hasFailed = true;
      }

      self.isWaiting = false;
    }),
    logOut: () => {
      removeToken();
      self.isLoggedIn = false;
      // Redirects to login page
      getParent(self).router.openLogInPage();
    },
  }));
