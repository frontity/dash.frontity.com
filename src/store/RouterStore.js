import { types } from 'mobx-state-tree';

export default types
  .model({
    page: 'sites',
    selectedSiteId: '',
  })
  .views(self => ({
    get currentUrl() {
      switch (self.page) {
        case 'sites':
          return '/';
        case 'new-site':
          return '/new-site';
        case 'site':
          return '/site/' + self.selectedSiteId;
        case 'login':
          return '/login';
        case 'signup':
          return '/signup';
        default:
          return '/404';
      }
    },
  }))
  .actions(self => ({
    openSitesPage() {
      self.page = 'sites';
      self.selectedSiteId = '';
    },
    openNewSitePage() {
      self.page = 'new-site';
      self.selectedSiteId = '';
    },
    openSitePage(site) {
      self.page = 'site';
      self.selectedSiteId = site.id;
    },
    openSitePageById(id) {
      self.page = 'site';
      self.selectedSiteId = id;
    },
    openLogInPage() {
      self.page = 'login';
      self.selectedSiteId = '';
    },
    openSignUpPage() {
      self.page = 'signup';
      self.selectedSiteId = '';
    },
  }));
