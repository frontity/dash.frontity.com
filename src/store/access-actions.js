import { flow } from 'mobx-state-tree';
import client, { setToken, removeToken } from '../graphql/client';

export const volatile = () => ({
  isSignUpModalOpen: false,
});

export const actions = self => ({
  logIn: flow(function* logIn({ value: emailAndPass }) {
    try {
      const { authenticateUser } = yield client.request(
        `
          mutation logIn($email: String!, $password: String!) {
            authenticateUser(email: $email, password: $password) {
              token
            }
          }
          `,
        emailAndPass,
      );
      setToken(authenticateUser.token);
      yield self.updateUser(); // Logged in User
    } catch (error) {
      console.error(error);
    }
  }),
  logOut: flow(function* logOut() {
    removeToken();
    yield self.updateUser(); // 'null'
  }),
  signUp: flow(function* signUp({ value: nameEmailAndPass }) {
    try {
      const { signupUser } = yield client.request(
        `
          mutation signUp($name: String!, $email: String!, $password: String!) {
            signupUser(name: $name, email: $email, password: $password) {
              token
            }
          }
        `,
        nameEmailAndPass,
      );
      setToken(signupUser.token);
      yield self.updateUser(); // Signed up User
      self.closeSignUpModal();
    } catch (error) {
      console.error(error);
    }
  }),
  openSignUpModal: () => (self.isSignUpModalOpen = true),
  closeSignUpModal: () => (self.isSignUpModalOpen = false),
});
