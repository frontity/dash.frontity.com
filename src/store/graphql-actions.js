import { flow, getType, recordPatches } from 'mobx-state-tree';
import client from '../graphql/client';

const updateQuery = async (node, data) => {
  const type = getType(node).name;
  const keys = Object.keys(data);
  const response = await client.request(
    `
    mutation update($id: ID!, ${keys.map(k => `$${k}: String`).join(', ')}) {
      update${type}(
        id: $id
        ${keys.map(k => `${k}: $${k}`).join(', ')}
      ) {
        id
      }
    }
  `,
    { id: node.id, ...data },
  );
  return response[`update${type}`];
};

export default self => ({
  update: flow(function* update(data) {
    const patches = recordPatches(self);
    try {
      Object.assign(self, data);
      patches.stop(self);
      yield updateQuery(self, data);
    } catch (error) {
      console.error(error);
      patches.undo(self);
    }
  }),
});
