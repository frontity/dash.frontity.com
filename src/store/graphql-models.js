import { types } from 'mobx-state-tree';

export const Package = types.model('Package', {
  id: types.identifier,
  name: types.string,
  namespace: types.string,
});

export const Setting = types.model('Setting', {
  id: types.identifier,
  active: types.boolean,
  package: types.reference(Package),
  data: types.frozen(),
});

export const Site = types.model('Site', {
  id: types.identifier,
  siteId: types.string,
  url: types.string,
  settings: types.array(Setting),
});

export const User = types.model('User', {
  id: types.identifier,
  name: types.string,
  email: types.string,
  sites: types.array(Site),
});
