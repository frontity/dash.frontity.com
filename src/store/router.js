import { reaction } from 'mobx';
import route from 'path-match';

export const createRouter = routes => {
  const matchers = Object.keys(routes).map(path => [
    route()(path),
    routes[path],
  ]);
  return function(path) {
    return matchers.some(([matcher, f]) => {
      const result = matcher(path);
      if (result === false) return false;
      f(result);
      return true;
    });
  };
};

export const initRouter = store => {
  reaction(
    () => store.router.currentUrl,
    path => {
      if (window.location.pathname !== path)
        window.history.pushState(null, null, path);
    },
  );

  const router = createRouter({
    '/new-site': store.router.openNewSitePage,
    '/site/:siteId': ({ siteId }) => store.router.openSitePageById(siteId),
    '/login': store.router.openLogInPage,
    '/signup': store.router.openSignUpPage,
    '/': store.router.openSitesPage,
  });

  window.onpopstate = function historyChange({ type }) {
    if (type === 'popstate') router(window.location.pathname);
  };

  router(window.location.pathname);
};
