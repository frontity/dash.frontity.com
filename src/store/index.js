import { types, flow } from 'mobx-state-tree';
import { initRouter } from './router';
import RouterStore from './router-store';
import * as access from './access-actions';
import { User } from './graphql-models';
import client from '../graphql/client';

const Store = types
  .model('Dashboard', {
    user: types.maybeNull(User),
    router: types.optional(RouterStore, {}),
  })
  .volatile(() => ({
    isWaitingForUser: true,
  }))
  .views(self => ({
    get isLoggedIn() {
      return !!self.user;
    },
  }))
  .actions(self => ({
    updateUser: flow(function* updateUser() {
      self.isWaitingForUser = true;
      const { user } = yield client.request(
        `
          query getUser {
            user {
              id
              name
              email
              sites {
                id
                siteId
                url
                settings {
                  id
                  active
                  data
                  package {
                    id
                    name
                    namespace
                  }
                }
              }
            }
          }
        `,
      );
      self.user = user;
      self.isWaitingForUser = false;
    }),
    afterCreate: () => self.updateUser(),
  }))
  .volatile(access.volatile)
  .actions(access.actions);

const store = Store.create({});
initRouter(store);

export default store;
