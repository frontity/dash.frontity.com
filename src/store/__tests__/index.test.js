import store from '..';

it('has the right properties', () => {
  expect(store).toMatchSnapshot();
});
