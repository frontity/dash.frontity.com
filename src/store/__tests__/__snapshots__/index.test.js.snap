// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has the right properties 1`] = `
Object {
  "router": Object {
    "page": "sites",
    "selectedSiteId": "",
  },
  "user": null,
}
`;
